docs(manifest): Improvements to `id` page (#36020)
mdn/content: files/en-us/web/manifest/id/index.md

{{QuickLinksWithSubpages("/en-US/docs/Web/Manifest")}}

> The `id` manifest member is used to represent the identity of your web application.

Author: Good catch, the same idea is getting repeated in the second sentence. I guess I picked that up from the spec, which talks about "id represents identity" and then "identity is used by user agents...".

I also noticed that I've used "represents" here, but on other pages I've been trying to stick with "...manifest member is used to specify..." to make clear it's something developers need to add.

So I'm thinking of going with this version (and dropping the second sentence):

> The `id` manifest member is used to specify a unique identifier for your web application.

> The `id` manifest member is used to represent the identity of your web application.

Reviewer: Perfect!

It is used by browsers to uniquely identify your application universally.

## Syntax

```json-nolint
"id": "com.example.myapp"
```

### Values

- `id`
  - : A string that takes the form of a URL.
    The URL must be same-origin with the [`start_url`](/en-US/docs/Web/Manifest/start_url) of your web app.
    If `id` is a relative URL, it is resolved using the origin of `start_url`. Any fragment in the `id` is always ignored.
    If `id` is not specified or the value is invalid in any way (such as not a string, not a valid URL, not same-origin with `start_url`), the `start_url` value is used.

## Description

> The `id` manifest member serves as a unique identifier for your web app. It allows browsers to distinguish between different applications, even if they are served from the same URL. This includes the following implications:

Reviewer: ```suggestion
The `id` manifest member serves as a unique identifier for your web app. It allows browsers to distinguish between different applications, even if they are served from the same URL:
```

Author: Simplified this even further to:
```suggestion
The `id` manifest member serves as a unique identifier for your web app. It allows browsers to distinguish between different applications:
```

Reviewer: Yup, this works well. The bullets handle the "from the same URL" part.

> - When a browser comes across an app manifest with an `id` that does not correspond to an already installed application, it treats that manifest as a description of a distinct application, even if it is served from the same URL as another application.

Reviewer: ```suggestion
- When a browser comes across an app manifest with an `id` that does not correspond to an already installed application, it treats that manifest as a definition of a distinct application, even if it is served from the same URL as another application.
```

Author: I'm inclined to us "description" in this context because that's what the spec uses.

I also think it might be more appropriate to use "description" because the manifest file provides information about the web app rather than strictly defining it.

- When a browser comes across an app manifest with an `id` that matches the identity of an already installed app, it treats the new manifest as a replacement for the existing app's manifest, even if the app is served from a different URL than the one installed previously.

> [!NOTE]
> While the `id` is processed like a URL, it does not point to a resource that can be accessed, so it is not required to be within the app's [scope](/en-US/docs/Web/Manifest/scope).

The `id` can also be used by services that collect lists of web apps to uniquely identify applications.

A few key points to remember while using the `id` member:

> - Use a leading `/` to be explicit that the `id` is a root-relative URL path.

Reviewer: ```suggestion
- Use a leading `/` to specify that the `id` is a root-relative URL path.
```

- Since `id` is resolved against `start_url`'s origin, `id` values such as `../foo`, `foo`, `/foo`, and `./foo` all resolve to the same identifier relative to the origin. For example, if `start_url` is `https://example.com/app/`, all these `id` values will resolve to `https://example.com/foo/`.
- Standard URL encoding and decoding rules apply when resolving the `id` value.

> - Fragments in the `id` are always removed during processing.

Reviewer: ```suggestion
- Fragments in the `id` are removed during processing.
```

- Query parameters in the `id` are preserved and included in the final resolved identifier.

## Examples

### Creating a distinct app version

Suppose you create a web app with the following manifest:

```json
{
  "name": "My Weather Application",
  "id": "https://example.com/weatherapp/v1",
  "start_url": "https://example.com/app"
}
```

If you later create another version of this app with significant changes and want it to be treated as a different app, you can add the manifest as:

```json
{
  "name": "My Weather Application",
  "id": "https://example.com/weatherapp/v2",
  "start_url": "https://example.com/app"
}
```

> In this case, even though both manifest files are served from the same URL, browsers will treat the new manifest as a description of a distinct application because the `id` is different. As a result, users can have both versions installed simultaneously.

Reviewer: ```suggestion
In this case, even though both manifest files are served from the same URL, browsers will treat the new manifest as a definition of a distinct application because the `id` is different. As a result, users can have both versions installed simultaneously.
```

### Updating an existing app

Consider a scenario where you deploy a web app with the following manifest:

```json
{
  "name": "My Weather Application",
  "id": "https://example.com/weatherapp/",
  "start_url": "https://old-domain.com/app"
}
```

However, you later decide to move the app to a different domain. You would then update the manifest as follows:

```json
{
  "name": "My Weather Application",
  "id": "https://example.com/weatherapp/",
  "start_url": "https://new-domain.com/app"
}
```

Browsers will treat this new manifest as an update to the existing app because the `id` values match. In this case, users will receive an update to their existing app, rather than being prompted to install a new app.

### Understanding `id` resolution

> Assume that the manifest URL for your app is `https://example.com/app/manifest.json`, and the `start_url` is defined as `https://example.com/app/home`. The following table demonstrates how different `id` values in the manifest will be resolved:

Reviewer: ```suggestion
Assume that the manifest URL for your app is `https://example.com/my-app/manifest.json`, and the `start_url` is defined as `https://example.com/my-app/home`. The following table demonstrates how different `id` values in the manifest will be resolved:
```

Reviewer: Also, is the manifest URL relevant to any of these examples?

Author: Nope, you're right, the manifest URL is not needed here. Thanks for catching the URL typo :)

| `id` in manifest              | Resolved `id`                      | Explanation                                                                                                                     |
| ----------------------------- | ---------------------------------- | ------------------------------------------------------------------------------------------------------------------------------- |
| undefined                     | `https://example.com/my-app/home`  | Defaults to `start_url`                                                                                                         |

> | undefined                     | `https://example.com/my-app/`      | Assuming `start_url` is `https://example.com/my-app/#here` for this example; `id` defaults to `start_url` with fragment removed |

Reviewer: This one comes off as a bit weird and jarring because in the previous paragraph, you already say `start_url` is something different. Maybe don't put this one in the table, and instead describe it in a separate paragraph below the table?

Author: Removed it from the table. Added it to the "fragments are removed" point in the "Description":

> - Fragments in the `id` are removed during processing. For example, if `id` is set to `foo#bar`, it will be resolved as `foo`. Similarly, if `id` is undefined and the `start_url` is `https://example.com/app/#home`, `id` will resolve to `https://example.com/app/`.

Reviewer: Great stuff!
